src/docfx/lib/git/GitCommitProvider.cs

```diff
@@ -1,412 +1,74 @@
-                        singleParent = true;
-                        break;
-                    }
-                }
-
-                if (!singleParent)
-                {
-                    // Otherwise, follow all parents.
-                    for (var i = 0; i < parentCount; i++)
-                    {
-                        add = true;
-                        commitsToFollow.Add((commit.Parents[i], parentBlobs[i]));
-                    }
-                }
-
-                if (add)
-                {
-                    result.Add(commit);
-                }
-            }
-
-            if (updateCache)
-            {
-                lock (commitCache)
-                {
-                    _cacheUpdated = true;
-                    commitCache.Add((headCommit.Sha.a, headBlob), (result.Select(c => c.Sha.a).ToArray(), 0));
-                }
-            }
-
-            return result.Select(c => c.GitCommit).ToList();
-        }
-
-        public Task SaveCache()
-        {
-            if (!_cacheUpdated || string.IsNullOrEmpty(_cacheFilePath))
-            {
-                return Task.CompletedTask;
-            }
-
-            PathUtility.CreateDirectoryFromFilePath(_cacheFilePath);
-
-            return ProcessUtility.WriteFile(_cacheFilePath, stream =>
-            {
-                using (var writer = new BinaryWriter(stream))
-                {
-                    // Create a snapshot of commit cache to ensure count and items matches.
-                    var commitCache = _commitCache.ToList();
-
-                    writer.Write(commitCache.Count);
-                    foreach (var (file, value) in commitCache)
-                    {
-                        lock (value)
-                        {
-                            writer.Write(file);
-                            writer.Write(Math.Min(value.Count, MaxCommitCacheCountPerFile));
-
-                            var lruValues = value.OrderBy(pair => pair.Value.lruOrder).Take(MaxCommitCacheCountPerFile);
-
-                            foreach (var ((commit, blob), (commitHistory, _)) in lruValues)
-                            {
-                                writer.Write(commit);
-                                writer.Write(blob);
-                                writer.Write(commitHistory.Length);
-
-                                foreach (var sha in commitHistory)
-                                {
-                                    writer.Write(sha);
-                                }
-                            }
-                        }
-                    }
-                }
-            });
         }
 
         public void Dispose()
         {
-            var repo = Interlocked.Exchange(ref _repo, IntPtr.Zero);
-            if (repo != IntPtr.Zero)
+            foreach (var value in _commitProviders.Values)
             {
-                git_repository_free(_repo);
+                value.Value.Result.provider.Dispose();
```

Comment:
There is probably no need to evaluate `Lazy` if it `IsValueCreated` is false.
Similarly, there is no need to evaluate Task `Result` if the Task isn't completed. Task is a little more complicated, in theory, task could be in running state when dispose if called, and if we choose to not explicitly dispose running tasks, we need to rely on FileCommitProvider finalizers to do the disposable. Unfortunately, `FileCommitProvider` currently do not have a finalizer, so we may need to add that as well.

Reply:
great comment, learned

---
In reply to: [237937253](https://github.com/dotnet/docfx/pull/3705#discussion_r237937253) [](ancestors = 237937253)